Initial Class structure for gift entry forms

callen-sfdo opened the pull request:
This is the first set of classes for the template structure.  I have not included the Data Import CMT wrappers as part of this, since I decided to include only references to them, not embed them within the class.

I think we may still need the wrappers though, but I want to think about it and discuss more.

Let me know if you have any questions.


# Critical Changes

# Changes

# Issues Closed

# New Metadata

Apex Classes:
* FORM_Element.cls
* FORM_Layout.cls
* FORM_Section.cls
* FORM_ServiceGiftEntry.cls
* FORM_ServiceGiftEntryTest.cls
* FORM_Template.cls

# Deleted Metadata

ptewson-sfdo commented on src/classes/FormServiceGiftEntry.cls:
```diff
@@ -0,0 +1,102 @@
+    public static FormTemplate retrieveFormTemplate(Id templateId) {
+        FormTemplate template;
+
+        Form_Template__c templateObj = [SELECT Id, Name, Format_Version__c, Template_JSON__c 
```
Are we missing the Form_Template__c object in this PR?

callen-sfdo replied:
Yep missed that, the object is now in this branch.

ptewson-sfdo commented on src/classes/FormServiceGiftEntryTest.cls:
```diff
@@ -0,0 +1,91 @@
+                                        'True',
+                                        null,
+                                        'CustomLabel',
+                                        new DataImportFieldMapping[]{});
```
```suggestion
                                        new String[]{});
```

ptewson-sfdo commented on src/classes/FormServiceGiftEntryTest.cls:
```diff
@@ -0,0 +1,91 @@
+        FormElement widget = new FormElement('FormWidget',
+                                            'DisplayRule',
+                                            'ComponentName',
+                                            new DataImportFieldMapping[]{});
```
```suggestion
                                            new String[]{});
```

vmartinov-sforg commented on src/classes/FORM_Layout.cls:
```diff
@@ -0,0 +1,48 @@
+public with sharing class FORM_Layout {
+
+    @AuraEnabled public String fieldMappingSetDevName;
+    @AuraEnabled public String version;
```
Is version an attribute of the Form Template or the form Layout?

callen-sfdo replied:
See above, I thought we had discussed having it at the layout level, but I am good moving it up to the template if that is the consensus.

callen-sfdo commented on src/classes/FORM_ServiceGiftEntry.cls:
```diff
@@ -0,0 +1,102 @@
+    public static FORM_Template retrieveDefaultFormTemplate() {
+        FORM_Template template;
+        // TBD: determine how default templates will be tracked for SGE.
+        Form_Template__c templateObj = [SELECT Id, Name, Format_Version__c, Template_JSON__c 
```
This is just a placeholder for future logic, I just wanted to add a stub for that method here first.

callen-sfdo commented on src/classes/FORM_Template.cls:
```diff
@@ -0,0 +1,46 @@
+    @AuraEnabled public String name;
+    @AuraEnabled public FORM_Layout layout;
+
+    public FORM_Template(String name, FORM_Layout layout) {
```
Yes to the description, just forgot to add it. For the version we had discussed whether it made sense to have it at the layout or the template level since the layout could theoretically be re-used independently of the template (which I thought was why we broke out the layout as a separate class to begin with) @ptewson-sfdo do you have an opinion here?

callen-sfdo commented on src/objects/Form_Template__c.object:
```diff
@@ -0,0 +1,187 @@
+    </fields>
+    <label>Form Template</label>
+    <nameField>
+        <label>Form Template Name</label>
```
yes, I forgot to add it, I will do so now.

callen-sfdo commented on src/objects/Form_Template__c.object:
```diff
@@ -0,0 +1,187 @@
+        <inlineHelpText>If there are changes made to the format/structure of the FormTemplate class or child classes this number indicates the version of the instance stored on this record.</inlineHelpText>
+        <label>Format Version</label>
+        <length>40</length>
+        <required>false</required>
```
I'll make it and the template JSON required.

callen-sfdo commented:
@ptewson-sfdo @vmartinov-sforg I've added a Description field to the Form Template object, and added it to the FORM_Template class.  I also moved version to the FORM_Template level, which I think makes sense if we are storing it on the Form_Template__c object.  If we want to add another version for the layout itself, we can do so at a later date.
